### PR TITLE
Refactor public interfaces of some low-level driver code for the Nonin sensor

### PR DESCRIPTION
**Pull request by rohanpurohit**
This PR includes:
- remove enums that were not used
- Reorder enums, consts to Types.h
- Rename `SignalAmplitude` to `SignalQuality`
- Rename `Pleth` to `PlethPulseAmplitudes`
- Rename `PacketMeasurements` to `Sample`
- Rename `StatusByteStruct` to `Status`
- Reorder `StatusByteStruct` fields as per datasheet
- Reorder fields in Sample as per datasheet (pg 6, Serial Data format (2) table)
- Rename `heart_rate` to `hr`
- Rename `spo2_d_beat` to `spo2_b_b`
- Rename `spo2_d_fast` to `spo2_fast` (as per datasheet)
- Rename `nonin_oem_revision` to `firmware_revision`
- Remove unnecessary code comments
- Use FrameBufferStatus for FrameBuffer instead of BufferStatus
- Rename `get_6bit_data` to `get_spo2_data` as it essentialy only gets SpO2 data from the packet
- Rename `get_9bit_data` to `get_hr_data`
- Rename `mask_6bit` to `spo2_missing_data` which is equal to `0x7F` or 127, this is an indicator for missing SpO2 data, we just ignore it if we get these bytes
- Rename `mask_9bit` to `hr_missing_data` which is `0x01FF` or 511 this is an indicator for missing HR data, we just ignore it if we get these bytes
- Rename `FrameInputStatus` `available` enum to `output_ready`
- Rename `FrameInputStatus` `framing_error` enum to `checksum_failed`
- Rename `PacketStatus` `framing_error` enum to `checksum_failed`
- Rename `PacketStatus` `missed_data` enum to `frame_loss`
- Rename `PacketInputStatus` `missed_data` enum to `frame_loss`
- Rename `PacketStatus` `available` enum to `ok`

TODO: check for other changes

### CODE REVIEW

**Comment by ethanjli** [comment]
Partial review of the existing Nonin OEM III code:

- Types.h:
    - Should we move enum class statuses for Receiver classes specific to those classes (e.g. "FrameInputStatus") into those classes (e.g. "FrameReceiver::InputStatus")?
    - enum class PacketStatus:
        - What does "not_available" mean? Presumably any value other than "available" should imply "not available".
        - What's the difference between "framing_error" and "missed_data"?
    - enum class SignalAmplitude:
        - Should this be renamed to SignalQuality?
    - struct StatusByteStruct:
        - Should this be renamed to FrameStatus or even just Status?
    - using StatusByteError:
        - Should this be renamed to FrameStatuses or even just Statuses?
        - Is this just dead code?
    - using Pleth:
        - Should this be renamed to PlethPulseAmplitudes?
    - struct PacketMeasurements:
        - Should this be renamed to Sample?
        - Should the fields in here be reordered to match the order of items on page 8 of the datasheet? (e.g. hr_d should be after spo2)
        - Should "heart_rate" in the field names be renamed to "hr" for consistency with the datasheet's names?
        - uint8_t spo2_d_beat:
            - Should this be renamed to spo2_b_b for consistency with the datasheet's name?
        - uint8_t nonin_oem_revision:
            - Should this be renamed to firmware_revision to match the description in the datasheet?
            - Should our Sensor class check that this value should be within some allowed range?
        - The datasheet says that spo2 = 127 and hr = 511 as sentinel values when those quantities cannot be computed. How are we reporting missing data here? Should we add bool fields has_spo2 and has_hr, or is that already reported in some other way?
        - std::array<bool, packet_size> ...:
            - Should we make a `using Flags = std::array<bool, packet_size>` and then just use that for all these declarations?
    - static const uint8_t mask_*:
        - Since these are single-byte masks, we should use binary notation (and we can use the digit grouping symbol `'` for readability)
- FrameBuffer.h
    - Frame frame_buffer_:
        - The doxygen comment for the `output` method has an unclear description which doesn't describe what it actually does
        - Should this be renamed to `frame_`?
    - uint8_t received_length:
        - This should be a `size_t` since it's used for array indexing (and then we can remove a static_cast in the cpp file)
- FrameBuffer.cpp
    - There are too many useless comments (e.g. `/* Increment the frame buffer index */`), it adds too much noise to code comments.
    - `received_length_++`:
        - In general we should use the prefix `++` and `--` rather than the postfix `++` and `--`: e.g. `++received_length_`
    - `BufferStatus input()`, `BufferStatus output()`:
        - The way these methods use `BufferStatus::partial` abuses the semantics for `BufferStatus`, where `partial` means "only a partial buffer operation was possible" - this is a safety problem. `FrameBuffer` should instead have something more like `ok` and `waiting`, and maybe `empty` and full` if needed, or else we should redesign the interface and separation of responsibilities between `FrameBuffer` and `BufferReceiver` to remove the need for a `partial` or `waiting` status code from `FrameBuffer`. Or we could just have `ok` and `waiting`, and also `input_overwritten` like in `ChunkSplitter`.
    - shift_left:
        - We should instead do `if (received_length_ == 0) return;`

File: firmware/ventilator-controller-stm32/Core/Inc/Pufferfish/Driver/Serial/Nonin/Types.h
Comment (ethanjli) [inline firmware/ventilator-controller-stm32/Core/Inc/Pufferfish/Driver/Serial/Nonin/Types.h]: It makes sense to me that `frame_max_size` and `packet_size` belong in Types.h, because they're configuration parameters which we took directly from the datasheet. However, maybe the type definitions of `Frame` and `Packet` can just live in the FrameReceiver.h/PacketReceiver.h/etc. files, since the specific choice of data structure is an implementation detail which doesn't come from the datasheet and isn't needed by the outside world (which only needs to use `Sample).

By this logic, we should still define `Flags`, `PlethPulseAmplitudes`, etc., in Types.h, because they're part of `Sample` which is used by the outside world.

File: firmware/ventilator-controller-stm32/Core/Inc/Pufferfish/Driver/Serial/Nonin/Types.h
Comment (ethanjli) [inline firmware/ventilator-controller-stm32/Core/Inc/Pufferfish/Driver/Serial/Nonin/Types.h:41]: Let's also add a code comment describing which Serial format we're using from the Nonin OEM III datasheet (I think it's format 2) for our low-level driver, including the URL of the datasheet.

File: firmware/ventilator-controller-stm32/Core/Inc/Pufferfish/Driver/Serial/Nonin/Types.h
Comment (ethanjli) [inline firmware/ventilator-controller-stm32/Core/Inc/Pufferfish/Driver/Serial/Nonin/Types.h]: Let's wrap these definitions in an internal namespace such as `StatusMasks`. Since these are implementation details which just come from the datasheet, it's fine for these definitions to be in `Driver::Serial::Nonin::StatusMasks` rather than being exported as part of the "public" interface of the file, which is `Driver::Serial::Nonin`.

File: firmware/ventilator-controller-stm32/Core/Inc/Pufferfish/Driver/Serial/Nonin/Types.h
Comment (ethanjli) [inline firmware/ventilator-controller-stm32/Core/Inc/Pufferfish/Driver/Serial/Nonin/Types.h]: Let's wrap these definitions in an internal namespace such as `ByteIndex`. This way we can also remove the `_index` postfix from each variable name.

File: firmware/ventilator-controller-stm32/Core/Inc/Pufferfish/Driver/Serial/Nonin/FrameBuffer.h
Comment (ethanjli) [inline firmware/ventilator-controller-stm32/Core/Inc/Pufferfish/Driver/Serial/Nonin/FrameBuffer.h:41]: Here is a comparison of the return code semantics of our two buffer classes:
- `Util::Containers::RingBuffer::push` returns `full` if there is no more space available to push bytes, so that the `push` could not successfully complete. Similarly, `Driver::Serial::Nonin::FrameBuffer::input` returns `full` if the input if there is no more space available to push bytes, so that the `input` operation could not successfully complete - which is what we want.
- `RingBuffer::push` returns `ok` if there more space available to push bytes. We should change `FrameBuffer::input` to have the same semantics; right now it returns `waiting` instead.
- `RingBuffer::push` doesn't have any concept of an output being ready or of the caller needing to stop adding bytes in order to process the ones already in the bufer. So `FrameBuffer::input` should have a new return code with a different name to indicate this; right now it returns `ok` instead, which conflicts with the semantics of `RingBuffer::push`'s `ok` return code.
- `RingBuffer::pop` returns `empty` if there are no bytes available to pop, so that the `pop` could not successfully complete. `FrameBuffer::output` should return some error code if there is not a complete buffer to inspect, so that the `output` could not successfully complete; right now it returns `waiting`, which is fine.
- `RingBuffer::pop` returns `ok` if there is a byte available to push. `RingBuffer::output` should return `ok` if there's a complete frame to inspect; right now it returns `ok`, which is what we want.

Either we should have two separate enums for `InputStatus` and `OutputStatus` (which is what I've done in the `Protocols::Transport` classes) or we should have a single enum for `FrameBufferStatus` but be more descriptive in the `@return` code comment about exactly which error codes it might return and what the semantics of it are (which is what I've done in RingBuffer.h, see [this example](https://github.com/pez-globo/pufferfish-software/blob/d386a033885c6d89196643199d5f374c170ebfe7/firmware/ventilator-controller-stm32/Core/Inc/Pufferfish/Util/Containers/RingBuffer.h#L79)).

Since this renaming is slightly more complicated (i.e. we're changing the meaning of `ok`, which is already being used elsewhere in the code), we could save it for the next PR - it's up to you.

File: firmware/ventilator-controller-stm32/Core/Src/Pufferfish/Driver/Serial/Nonin/Device.cpp
Comment (ethanjli) [inline firmware/ventilator-controller-stm32/Core/Src/Pufferfish/Driver/Serial/Nonin/Device.cpp]: This made me nervous because I initially thought there were error codes in `PacketOutputStatus` and we were just ignoring them without a `TODO` comment; it turns out that right now `PacketOutputStatus` only has `ok` and `waiting`, so it doesn't have error codes.

If `clang-tidy` allows it, let's convert this to a two-case `switch` statement handling `ok` and `waiting`, so that we should get a clang-tidy warning about an unhandled enum in the `switch` if someone adds another code (e.g. an error code which needs to be handled) to `PacketOutputStatus` in the future.

File: firmware/ventilator-controller-stm32/Core/Inc/Pufferfish/Driver/Serial/Nonin/FrameReceiver.h
Comment (ethanjli) [inline firmware/ventilator-controller-stm32/Core/Inc/Pufferfish/Driver/Serial/Nonin/FrameReceiver.h:32]: [probably for next PR] We should probably split this up into two error codes: `invalid_header`/`invalid_start_of_frame` (i.e. `new_frame[0] != 0x01 || (new_frame[1] & mask_start_of_frame) != mask_start_of_frame` in `validate_frame`) and `invalid_checksum`.

File: firmware/ventilator-controller-stm32/Core/Src/Pufferfish/Driver/Serial/Nonin/FrameReceiver.cpp
Comment (ethanjli) [inline firmware/ventilator-controller-stm32/Core/Src/Pufferfish/Driver/Serial/Nonin/FrameReceiver.cpp:31]: [probably for the next PR] From looking at the code, this looks like it should actually be `validate_start_of_packet`? It's weird to me that `FrameReceiver` expects to know anything about the structure of packets, such as whether a frame is at the start of a packet.

File: firmware/ventilator-controller-stm32/Core/Src/Pufferfish/Driver/Serial/Nonin/FrameReceiver.cpp
Comment (ethanjli) [inline firmware/ventilator-controller-stm32/Core/Src/Pufferfish/Driver/Serial/Nonin/FrameReceiver.cpp:46]: [probably for next PR] Right now `validate_frame` has an unnecessary nested `if` statement structure. We should probably just split up `validate_frame` into two functions `validate_frame_header` and `validate_frame_checksum` and use them directly in `update_frame_buffer`.

File: firmware/ventilator-controller-stm32/Core/Inc/Pufferfish/Driver/Serial/Nonin/FrameReceiver.h
Comment (ethanjli) [inline firmware/ventilator-controller-stm32/Core/Inc/Pufferfish/Driver/Serial/Nonin/FrameReceiver.h:81]: [for the next PR] The code comment for `start_of_frame_status` is confusing to me because it's talking about packets (and also the descriptions were already run-on sentences with a structure I don't understand, before clang-format merged separate lines for `true` and `false`); but FrameReceiver.h should have no awareness of anything about packets. Maybe we can also make the variable slightly shorter and clearer by naming it `at_start_of_frame` or something.

File: firmware/ventilator-controller-stm32/Core/Src/Pufferfish/Driver/Serial/Nonin/FrameReceiver.cpp
Comment (ethanjli) [inline firmware/ventilator-controller-stm32/Core/Src/Pufferfish/Driver/Serial/Nonin/FrameReceiver.cpp:87]: [for next PR] I don't think we need to use `this->`, and in general we don't use `this->` in our codebase.

File: firmware/ventilator-controller-stm32/Core/Src/Pufferfish/Driver/Serial/Nonin/Sensor.cpp
Comment (ethanjli) [inline firmware/ventilator-controller-stm32/Core/Src/Pufferfish/Driver/Serial/Nonin/Sensor.cpp:33]: Do we need to add a TODO comment about handling the case where the `Device::output` method doesn't return `ok`? Right now it seems that `Sensor::output` returns `ok` regardless of what happens. Maybe just what we need to change in order to add POST checks for the Nonin sensor? (there's probably other things we'll need to change, too).

This is probably code I wrote in a hurry. If I had used a return-early structure (i.e. `status = device_.output(...); if (status !== ok) {return an error code or do nothing but have a TODO comment for dealing with this} `), this would have been easier to catch and understand.

File: firmware/ventilator-controller-stm32/Core/Inc/Pufferfish/Driver/Serial/Nonin/PacketReceiver.h
Comment (ethanjli) [inline firmware/ventilator-controller-stm32/Core/Inc/Pufferfish/Driver/Serial/Nonin/PacketReceiver.h]: Is `spo2_missing_data` a mask? If so, should `spo2_missing_data` be moved to be with the other masks in Types.h? Also, what does "spo2 missing data" mean? Would it be clear if we just named it to "spo2 missing"? I think "spo2 missing data" is too ambiguous because it could either be saying "There is something called SpO2 and it needs some data which is currently missing", "We are missing the data for SpO2", or "Here is some data describing why SpO2 is missing".

If `spo2_missing_data` is a non-mask value defined in the datasheet, perhaps we should move it to `Types.h` (maybe into a `SpecialConstants` or `ErrorConstants` or `Flags` internal namespace) and use the datasheet's format for representing the value, e.g. `static const uint8_t spo2_missing = 127; // Normally we'd use hex, but the datasheet lists it as 127 in decimal so we do the same thing here`.

File: firmware/ventilator-controller-stm32/Core/Inc/Pufferfish/Driver/Serial/Nonin/PacketReceiver.h
Comment (ethanjli) [inline firmware/ventilator-controller-stm32/Core/Inc/Pufferfish/Driver/Serial/Nonin/PacketReceiver.h]: If `hr_missing_data` is actually a value defined in the datasheet, perhaps we should move into Types.h together with the "SpO2 missing" flag

File: firmware/ventilator-controller-stm32/Core/Src/Pufferfish/Driver/Serial/Nonin/PacketReceiver.cpp
Comment (ethanjli) [inline firmware/ventilator-controller-stm32/Core/Src/Pufferfish/Driver/Serial/Nonin/PacketReceiver.cpp]: [for a future PR] To avoid copying `sensor_measurements.signal_perfusion[frame_index]` so many times, maybe we could first get a non-const ref as `SignalPerfusion &signal_perfusion = sensor_measurements.signal_perfusion[frame_index]` and then just assign to `signal_perfusion`.

File: firmware/ventilator-controller-stm32/Core/Src/Pufferfish/Driver/Serial/Nonin/PacketReceiver.cpp
Comment (ethanjli) [inline firmware/ventilator-controller-stm32/Core/Src/Pufferfish/Driver/Serial/Nonin/PacketReceiver.cpp]: [for a future PR] I wonder what `byte3 means` and why we use it so many times. Maybe we can find a more descriptive name, e.g. have a `ByteIndices` namespace in Types.h with constants for `header_byte`, `status_byte`, `pleth_byte`, `payload_byte`/`data_byte`, and `checksum_byte`.

File: firmware/ventilator-controller-stm32/Core/Src/Pufferfish/Driver/Serial/Nonin/PacketReceiver.cpp
Comment (ethanjli) [inline firmware/ventilator-controller-stm32/Core/Src/Pufferfish/Driver/Serial/Nonin/PacketReceiver.cpp:85]: Since this `if` statement is changing a lot of internal state (which is where our codebase is most error-prone), we should use descriptive naming and code comments to help communicate what's going on.

What does the expression being tested by the `if` statement actually mean? We can answer this question in the code by having a descriptive function name to replace `(frame[1] & 0x01U) == 0x01`.

What does this `if` statement try to achieve when the condition is met? We can answer this question in the code by adding a code comment which summarizes everything inside, as a replacement for the line-by-line code comments which take too much work to try to compile into a summary-level understanding.

File: firmware/ventilator-controller-stm32/Core/Src/Pufferfish/Driver/Serial/Nonin/PacketReceiver.cpp
Comment (ethanjli) [inline firmware/ventilator-controller-stm32/Core/Src/Pufferfish/Driver/Serial/Nonin/PacketReceiver.cpp]: This could be shortened to `++received_length`.

**Review by ethanjli — COMMENTED** [review]
I reviewed all sections of code touched by these changes; all of your changes look good. Most of these are suggestions for the next Nonin driver refactoring PR. Since there is some code changed which would not be trivially identified by compiler errors, later today I'll test the updated driver code with my Nonin sensor just to confirm that there aren't any obvious regressions in normal operation. Error-handling is complicated and needs more refactoring and possibly some deeper changes, so I won't test that in this PR.

All of the suggested changes for this PR in this review should be minor changes (anything which a compiler can't immediately and obviously catch as an error should be saved for a future PR; and anything about the quality of files besides FrameBuffer.h and Types.h is technically outside the scope of this PR), so if you add your responses to the records-keeping questions for open-source licensing, then I'll approve this PR for merging after very light testing.

**Comment by rohanpurohit** [comment]
1. This project is licensed under Apache License v2.0 for any software, and Solderpad Hardware License v2.1 for any hardware - do you agree that your contributions to this project will be under these licenses, too? Yes.
2. Were any of these contributions also part of work you did for an employer or a client? No.
3. Does this work include, or is it based on, any third-party work which you did not create? No.

**Review by ethanjli — APPROVED** [review]
I just tested the firmware together with my Nonin sensor, and everything seems to work fine in the normal case!

**Pull request merged**
